odl/trafos/fourier.py

```diff
@@ -1645,6 +1647,211 @@ def inverse(self):
+                self._is_sample_valid(sample) for sample in non_uniform_samples
+            ):
+            raise ValueError('One of the samples in `non_uniform_samples` is'
+              ' not a float between -0.5 and 0.5')
+
+    def _is_sample_valid(self, sample):
+        """Check whether a single sample is correct.
+
+        Parameters
+        ----------
+        sample : iterable
+            The sample whose location you want to check
+
+        Returns
+        -------
+        bool : Whether the sample contains only floats in [-0.5; 0.5[
+        """
+        floats = all(isinstance(coord, float) for coord in sample)
+        if floats:
+            normalized = all(-0.5 <= coord < 0.5 for coord in sample)
+        else:
+            normalized = False
+        return floats & normalized
+
+    def _normalize(self, x):
+        """Normalize the result of the non uniform FFT.
+
+        Parameters
+        ----------
+        x : `numpy.ndarray`
+            The results of the non uniform FFT
+
+        Returns
+        -------
+        out : `numpy.ndarray`
+            Normalized results of the non uniform FFT
+        """
+        out = x / np.sqrt(self.nfft.M)
+        return out
+
+    @property
+    def adjoint(self):
+        """Adjoint of this instance."""
+        if self.adjoint_class:
+            return self.adjoint_class(
+                shape=self.shape,
+                non_uniform_samples=self.non_uniform_samples,
+            )
+        raise NotImplementedError(
+            "Adjoint not implemented for this non-uniform fourier operator",
+        )
+
+
+class NonUniformFourierTransform(NonUniformFourierTransformBase):
+    """Forward Non uniform Fast Fourier Transform.
+    """
+    def __init__(self, shape, non_uniform_samples):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        shape : tuple
+            The dimensions of the data whose non uniform FFT you want to
+            compute
+        non_uniform_samples : iterable
+            List of the fourier space positions where the coefficients are
+            computed. Each position must be in [-0.5; 0.5[
+        """
+        super(NonUniformFourierTransform, self).__init__(
+            shape=shape,
+            non_uniform_samples=non_uniform_samples,
+            domain=discr_sequence_space(shape, dtype=np.complex128),
+            range=discr_sequence_space(
+                [len(non_uniform_samples)],
+                dtype=np.complex128,
+            ),
+        )
+        self.adjoint_class = NonUniformFourierTransformAdjoint
+
+    def _call(self, x):
+        """Compute the direct non uniform FFT.
+
+        Parameters
+        ----------
+        x : `numpy.ndarray`
+            The data whose non uniform FFT you want to compute
+
+        Returns
+        -------
+        out_normalized : `numpy.ndarray`
+            Result of the transform
+        """
+        self.nfft.f_hat = np.array(x)
```

Comment:
`x.asarray()` does not copy. If `nfft` destroys the input, then we need a copy.

Reply:
`nfft` doesn't destroy the input (just tested it).